Components/ImportService.php

```diff
@@ -544,4 +543,80 @@ private function getAst($categoryId, $query = '', $showOnlyConnectArticles = tru
 
         return $ast;
     }
+
+    /**
+     * @param array $articleIds
+     * @param $categoryId
+     */
+    private function unAssignArticlesFromCategory(array $articleIds, $categoryId)
+    {
+        $categories = [];
+        $categories[] = $categoryId;
+        $childCategories = $this->manager->getConnection()->executeQuery('SELECT id FROM s_categories WHERE path LIKE ?',
+            ["%|$categoryId|%"]);
+
+        while ($childCategory = $childCategories->fetchColumn()) {
+            $categories[] = (int) $childCategory;
+        }
+
+        $categoriesStatement = $this->manager->getConnection()->prepare('DELETE FROM s_articles_categories WHERE articleID IN (' . implode(', ', $articleIds) . ') AND categoryID IN (' . implode(', ', $categories) . ')');
+        $categoriesStatement->execute();
+
+        //Deletes all entries in s_articles_categories_ro that are not in the Path to a assigned category
+        //need the nested select because mysql got an error otherwise
+        //see: https://stackoverflow.com/questions/45494/mysql-error-1093-cant-specify-target-table-for-update-in-from-clause
+        $this->manager->getConnection()->executeQuery('
+                  DELETE FROM s_articles_categories_ro
+                  WHERE id NOT IN 
```

Comment:
I think NOT IN is too dangerous. In my shop I've 140K products. Most of them don't have any category, but usually every article has more than 1 category. So s_articles_categories_ro will contain more than 140K rows. Then 2 nested selects with 2 inner joins and LIKE is good enough. Actually why we need this query? As I said your already know articleIds and categoryIds.  then we can remove them one by one.

Reply:
The problem is in s_articles_categories_ro are all parent categories assigned to the articles stored
e.g 
i have categories tree like:
1:
    1.1:
        1.1.1
        1.1.2
    1.2
In s_articles_categories i have:
|articleID|categoryID|
           1 | 1.1.1
           1 | 1.2
this leads to s_articles_categories_ro also containing categories 1 and 1.1 for article 1